yetanotherco/aligned_layer
# Fix operator merkle root retrocompatibility

# This PR
Adds a second merkle root check. If the normal merkle root check returns false, it tries again with the old version of the SDK.
This is to ensure smooth operator upgradeability to v0.7.0

## Deploy
This changes should be checked out by operators (in v0.7.0)

- Operators who have not checked out, will accept tasks because they will be using the old version of the SDK.
- Operators who have updated to this version, will accept tasks because of the second check, using the old version of the SDK.
- After operators have checked to this version, we will be able to update our batcher to the v0.7.0 , and all new operators will be compatible because of the normal check, using the latest version of the SDK.

## Test
Note: you will need 2 repos to test this.

To test this works, first we must understand the fail.
The failure was given when the batcher was updated to v0.7.0 . This made old operators to not be compatible and not accept the merkle root.

### Retrocompatibility

1) Deploy all infra on `testnet` except Operator
2) Deploy Operator on this branch, ON ANOTHER REPO
- Don't forget to set `var switchBlockNumber = uint32(0)` in `operator.go`, `aggregator.go` and `subscriber.go` instances
- Also set `if block.number < 0`in `AlignedLayerServiceManager.sol`
3) Verify Operator receives new batches and verifies them correctly
- Note: Operator responds twice the event, once for each `NewBatchEvent` it detected (v2 and v3). This issue is solved in #1027

### Normal Compatibility 
1) Deploy all the network on this branch, and test normal flow

---
## Previous Version of this PR Description:

### Optional: test failure
- To test this failure case, you can check out to the latest stage version:
```
git checkout dd19e5f69b001ca4e040c706f04b72de43825d5f
```
- In here, it is very important to:
```
make deps
```
```
make install_aligned_compiling
```
- Then you must run anvil, aggregator and batcher:
```
make anvil_start_with_block_time
```
```
make aggregator_start
```
```
make batcher_start_local
```
- Once this is running, you can go to your other local repo, ON ANOTHER FOLDER, and checkout staging (what would currently be v0.7.0, without the current PR fix):
```
git checkout staging
```
- Here, you must rebuild ffi of this
```
make deps
```
- Then you can run operator
```
make operator_register_and_start
```
- And going back to your first repo, you can send tasks. Note the operator should reject them because of merkle root failed
```
make batcher_send_burst_groth16
```


### Test the fix
Having understood the fail case, we can now verify the fix is working:
- On one repo, checkout the latest version deployed on stage:
```
git checkout dd19e5f69b001ca4e040c706f04b72de43825d5f
```
- In here, it is very important to:
```
make deps
```
- Then you must run anvil, aggregator and batcher:
```
make anvil_start_with_block_time
```
```
make aggregator_start
```
```
make batcher_start_local
```
- Once this is running, you can go to your other local repo, ON ANOTHER FOLDER, and checkout this PR's branch
```
git checkout fix-operator-merkle_root-retrocompatibility
```
- Here, you must rebuild ffi of this
```
make deps
```
- Then you can run operator
```
make operator_register_and_start
```
- And going back to your first repo, you can send tasks.
```
make batcher_send_burst_groth16
```
- Note the operator should accept the merkle root, but only after trying with the old version:
<img width="679" alt="image" src="https://github.com/user-attachments/assets/db2defa8-3058-479a-8410-dd82100d2a18">

- We have verified the Operator retrocompatibility works.

### Test normal compatibility
Now you can test the normal compatibility. The idea of this is that after Operators have migrated to this version, the batcher must also change version.

For this you need to run the whole project on this branch.
```
git checkout fix-operator-merkle_root-retrocompatibility
```
```
make deps
```
```
make install_aligned_compiling
```
```
make anvil_start_with_block_time
```
```
make aggregator_start
```
```
make batcher_start_local
```
```
make operator_register_and_start
```
```
make batcher_send_burst_groth16
```
- It should accept the task as normally.
<img width="319" alt="image" src="https://github.com/user-attachments/assets/0fb08eed-1b20-4a10-8fc4-8f7bf9b301d1">
